cl/phase1/forkchoice/optimistic/optimistic_impl.go

```diff
@@ -0,0 +1,119 @@
+	if block.Body.ExecutionPayload == nil || *block.Body.ExecutionPayload == (cltypes.Eth1Block{}) {
+		return nil
+	}
+
+	root := block.StateRoot
+	parentRoot := block.ParentRoot
+	impl.opMutex.Lock()
+	defer impl.opMutex.Unlock()
+	if _, ok := impl.optimisticRoots[root]; ok {
+		// block already optimistically imported
+		return nil
+	}
+	blockNode := &opNode{
+		execBlockNum: block.Body.ExecutionPayload.BlockNumber,
+		parent:       parentRoot,
+		children:     []common.Hash{},
+	}
+	impl.optimisticRoots[root] = blockNode
+
+	// check if parent is already in the store
+	if _, ok := impl.optimisticRoots[parentRoot]; ok {
+		impl.optimisticRoots[parentRoot].children = append(impl.optimisticRoots[parentRoot].children, root)
+	}
+	return nil
+}
+
+func (impl *optimisticStoreImpl) ValidateBlock(block *cltypes.BeaconBlock) error {
+	// When a block transitions from NOT_VALIDATED -> VALID, all ancestors of the block MUST also transition
+	// from NOT_VALIDATED -> VALID. Such a block and any previously NOT_VALIDATED ancestors are no longer considered "optimistically imported".
+	if block.Body.ExecutionPayload == nil || *block.Body.ExecutionPayload == (cltypes.Eth1Block{}) {
+		return nil
+	}
+	blockNum := block.Body.ExecutionPayload.BlockNumber
+	impl.opMutex.Lock()
+	defer impl.opMutex.Unlock()
+	curRoot := block.StateRoot
+	for {
+		if node, ok := impl.optimisticRoots[curRoot]; ok {
+			// validate the block
+			// remove the block from the store
+			delete(impl.optimisticRoots, curRoot)
+			curRoot = node.parent
+		} else {
+			break
+		}
+	}
+	// and try to clean up all nodes with block number less than blockNum
+	toRemoves := []common.Hash{}
+	for root, node := range impl.optimisticRoots {
+		if node.execBlockNum < blockNum {
+			toRemoves = append(toRemoves, root)
+		}
+	}
+	for _, root := range toRemoves {
+		delete(impl.optimisticRoots, root)
+	}
+	return nil
+}
+
+func (impl *optimisticStoreImpl) InvalidateBlock(block *cltypes.BeaconBlock) error {
+	// When a block transitions from NOT_VALIDATED -> INVALIDATED, all descendants of the block MUST also transition
+	// from NOT_VALIDATED -> INVALIDATED.
+	if block.Body.ExecutionPayload == nil || *block.Body.ExecutionPayload == (cltypes.Eth1Block{}) {
+		return nil
+	}
+	impl.opMutex.Lock()
+	defer impl.opMutex.Unlock()
+	// start from the block to be invalidated, and remove all its descendants
+	toRemoves := []common.Hash{block.StateRoot}
+	for len(toRemoves) > 0 {
+		curRoot := toRemoves[0]
+		toRemoves = toRemoves[1:]
+		if node, ok := impl.optimisticRoots[curRoot]; ok {
+			// remove the invalidated block from the store
+			delete(impl.optimisticRoots, curRoot)
+			toRemoves = append(toRemoves, node.children...)
+		}
+	}
+	return nil
+}
+
+func (impl *optimisticStoreImpl) IsOptimistic(root common.Hash) bool {
+	if root == (common.Hash{}) {
+		return false
+	}
+	impl.opMutex.RLock()
+	defer impl.opMutex.RUnlock()
+	if _, ok := impl.optimisticRoots[root]; ok {
```

Comment:
If you delete (cleanup) a block root then it will be not optimistic anymore. I have a suggestion:

Clean if the slot of the block number is below `finalizedEpoch * slotsPerEpoch`, and if the root we get as a `slot <=  finalizedEpoch * slotsPerEpoch` then count it as `optimistic=true`.

Additionally add unit tests.

Reply:
Maybe you could pass `slot uint64` do and use the `syncedDataManager` for that

Reply:
I feel a bit cannot get your idea. Hmmmm maybe the reason is I don't really understand the block production flow. Here are a few questions:

1. `If you delete (cleanup) a block root then it will be not optimistic anymore`. But for now, current impl doesn't not clean up optimistic block from the optimistic candidates map, unless 1) itself or one of its descendant gets validated or 2) its block number less than largest validated block number ([here](https://github.com/ledgerwatch/erigon/pull/10179/files/17bfe570a9c93a865346ffde2106bd5087a72ec5#diff-e19af91c00493c6f2d22dadbcac820db8ec6f56864f409e35ce2448cdf519436R71)). Did I miss anything? or you advice me to use `slot` instead of `block number`?

2. According to your suggestion, you mean a block slot less than finalized checkpoint slot must be `Validated` ? That is, the payload must already be validated by execution node?
    - If so, then `slot <=  finalizedEpoch * slotsPerEpoch` should get optimistic status as `False`, right? And, in this case, i guess we don't need to do more things becasue the validated block and its ancestors should be removed  by returned validated status from `NewPayload()`
    - If not (maybe we are uncertain if finalized block is already checked by execution node), then keep the block in optimistic candidates map and don't need to remove it?

Reply:
Oh actually no I am blind. you are correct lol.

Reply:
let me test it

Reply:
Just write some unit tests perhaps